src/openapi-generator/helpers/action-controller.cr

```diff
@@ -0,0 +1,431 @@
+
+  # :nodoc:
+  CONTROLLER_RESPONSES = {} of String => ControllerResponsesValue
+  # :nodoc:
+  HASH_ITEM_REF = [] of {Int32, OpenAPI::Response}
+  # :nodoc:
+  TYPE_REF = [] of String
+  # :nodoc:
+  QP_LIST = {} of String => Array(OpenAPI::Parameter)
+  # :nodoc:
+  BODY_LIST = {} of String => {OpenAPI::RequestBody, Hash(String, OpenAPI::Schema)}
+
+  # Fetch a query parameter and register it in the OpenAPI operation related to the controller method.
+  #
+  # ```
+  # query_params "name", "A user name."
+  # ```
+  macro query_params(name, description, multiple = false, schema = nil, **args)
+    _query_params(
+      name: {{name}},
+      param: ::OpenAPI::Generator::Helpers::ActionController.init_openapi_parameter(
+        name: {{name}},
+        "in": "query",
+        required: true,
+        schema: {% if schema %}{{schema}}{% elsif multiple %}::OpenAPI::Schema.new(
+          type: "array",
+          items: ::OpenAPI::Schema.new(
+            type: "string"
+          )
+        ){% else %}::OpenAPI::Schema.new(
+          type: "string",
+        ){% end %},
+        description: {{description}},
+        {{**args}}
+      ),
+      required: true,
+      multiple: {{multiple}}
+    )
+  end
+
+  # Fetch an optional query parameter and register it in the OpenAPI operation related to the controller method.
+  #
+  # ```
+  # query_params? "name[]", "One or multiple user names. (optional)", multiple = true
+  # ```
+  macro query_params?(name, description, multiple = false, schema = nil, **args)
+    _query_params(
+      name: {{name}},
+      param: ::OpenAPI::Generator::Helpers::ActionController.init_openapi_parameter(
+        name: {{name}},
+        "in": "query",
+        required: false,
+        schema: {% if schema %}{{schema}}{% elsif multiple %}::OpenAPI::Schema.new(
+          type: "array",
+          items: ::OpenAPI::Schema.new(
+            type: "string"
+          )
+        ){% else %}::OpenAPI::Schema.new(
+          type: "string",
+        ){% end %},
+        description: {{description}},
+        {{**args}}
+      ),
+      required: false,
+      multiple: {{multiple}}
+    )
+  end
+
+  # :nodoc:
+  private macro _query_params(name, param, required = true, multiple = false)
+    {% qp_list = ::OpenAPI::Generator::Helpers::ActionController::QP_LIST %}
+      {% method_name = "#{@type}::#{@def.name}" %}
+      {% unless qp_list.keys.includes? method_name %}
+        {% qp_list[method_name] = [] of OpenAPI::Parameter %}
+      {% end %}
+      {% qp_list[method_name] << param %}
+      {% if multiple %}
+        %results = params.fetch_all({{name}})
+        {% if required %}
+          raise NilAssertionError.new if %results.size < 1
+        {% end %}
+        %results
+      {% elsif required %}
+        params[{{name}}]
+      {% else %}
+        params[{{name}}]?
```

Comment:
Could be a place to use the http-params-serializable lib here, and cast to the desired type.